File: server/connection.go
package hex

import (
	"fmt"
	"net"
	"net/http"
	"net/url"
	"path/filepath"
	"strings"
)

type ConnectionConfig struct {
	Name       string
	Server     string
	Publishing *PublishingConfig
}

type PublishingConfig struct {
	Provider string `json:"provider"`
	Root     string `json:"root,omitempty"`
	URL      string `json:"url,omitempty"`
}

type connectionDocument struct {
	Version       int               `json:"version"`
	Name          string            `json:"name"`
	Server        string            `json:"server"`
	SiteBaseURL   string            `json:"siteBaseURL"`
	CLIReleaseURL string            `json:"cliReleaseURL,omitempty"`
	Publishing    *PublishingConfig `json:"publishing,omitempty"`
	Capabilities  map[string]any    `json:"capabilities"`
}

func (s *Server) connectionConfig(w http.ResponseWriter, r *http.Request) {
	document, err := s.connectionSettings()
	if err != nil {
		writeServerError(w, err)
		return
	}

	w.Header().Set("Content-Disposition", `attachment; filename="hex-platform.json"`)
	writeJSON(w, http.StatusOK, document)
}

func (s *Server) connectionSettings() (connectionDocument, error) {
	connection := s.config.Connection
	if connection == nil {
		return connectionDocument{}, fmt.Errorf("platform connection settings are not configured")
	}
	if err := validateConnection(connection, s.config.SiteBaseURL); err != nil {
		return connectionDocument{}, fmt.Errorf("invalid platform connection settings: %w", err)
	}
	if _, err := s.cliReleaseURL(); err != nil {
		return connectionDocument{}, err
	}
	return connectionDocument{
		Version:       1,
		Name:          connection.Name,
		Server:        connection.Server,
		SiteBaseURL:   s.config.SiteBaseURL,
		CLIReleaseURL: s.config.CLIReleaseURL,
		Publishing:    connection.Publishing,
		Capabilities:  s.capabilityDescription(),
	}, nil
}

func validateConnection(connection *ConnectionConfig, siteBaseURL string) error {
	if strings.TrimSpace(connection.Name) == "" {
		return fmt.Errorf("platform name is required")
	}
	server, err := connectionOrigin(connection.Server)
	if err != nil {
		return err
	}
	if _, err := parseSiteBaseURL(siteBaseURL); err != nil {
		return err
	}
	if connection.Publishing == nil {
		return nil
	}

	publishing := connection.Publishing
	switch publishing.Provider {
	case "filesystem":
		if !loopbackHost(server.Hostname()) {
			return fmt.Errorf("filesystem publishing is only advertised by local platforms")
		}
		if !filepath.IsAbs(publishing.Root) || publishing.URL != "" {
			return fmt.Errorf("filesystem publishing requires an absolute root and no URL")
		}
	case "azure-files":
		destination, err := url.Parse(publishing.URL)
		if err != nil || destination.Scheme != "https" || destination.Host == "" || destination.User != nil || destination.RawQuery != "" || destination.Fragment != "" {
			return fmt.Errorf("Azure Files publishing requires an HTTPS URL without credentials or query parameters")
		}
		if destination.Path == "" || destination.Path == "/" || publishing.Root != "" {
			return fmt.Errorf("Azure Files publishing requires a share/prefix URL and no filesystem root")
		}
	default:
		return fmt.Errorf("unsupported publishing provider")
	}
	return nil
}

func connectionOrigin(value string) (*url.URL, error) {
	parsed, err := url.Parse(value)
	if err != nil {
		return nil, fmt.Errorf("invalid platform server URL")
	}
	validScheme := parsed.Scheme == "https" || (parsed.Scheme == "http" && loopbackHost(parsed.Hostname()))
	if !validScheme || parsed.Host == "" || parsed.User != nil || parsed.RawQuery != "" || parsed.Fragment != "" || (parsed.Path != "" && parsed.Path != "/") {
		return nil, fmt.Errorf("platform server must be an HTTPS origin, or HTTP on loopback")
	}
	return parsed, nil
}

func loopbackHost(host string) bool {
	return host == "localhost" || strings.HasSuffix(host, ".localhost") || net.ParseIP(host).IsLoopback()
}

File: packages/cli/src/connection.mjs
import { isIP } from "node:net";
import { isAbsolute, win32 } from "node:path";

export const connectionPath = "/api/hex/config";
export const maxConfigBytes = 64 * 1024;

export function isLocalHost(hostname) {
  const host = hostname.replace(/^\[|\]$/g, "");
  return (
    host === "localhost" ||
    host.endsWith(".localhost") ||
    host === "::1" ||
    (isIP(host) === 4 && host.startsWith("127."))
  );
}

export function platformOrigin(value) {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error("A platform URL is required");
  }
  const url = new URL(value);
  const validScheme =
    url.protocol === "https:" ||
    (url.protocol === "http:" && isLocalHost(url.hostname));
  if (
    !validScheme ||
    url.username ||
    url.password ||
    url.search ||
    url.hash ||
    url.pathname !== "/"
  ) {
    throw new Error(
      "Use an HTTPS platform origin, or HTTP on localhost, without credentials, paths or query parameters",
    );
  }
  return url.origin;
}

function object(value, description) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${description} must be an object`);
  }
}

function knownFields(value, allowed, description) {
  object(value, description);
  if (Object.keys(value).some((key) => !allowed.includes(key))) {
    throw new Error(
      `${description} contains unsupported fields; connection files must not contain credentials or commands`,
    );
  }
}

function validatePublishing(value, server) {
  if (value === undefined || value === null) {
    return undefined;
  }
  knownFields(value, ["provider", "root", "url"], "Publishing configuration");

  if (value.provider === "filesystem") {
    if (!isLocalHost(new URL(server).hostname)) {
      throw new Error(
        "Filesystem publishing is only supported for local platforms",
      );
    }
    if (
      typeof value.root !== "string" ||
      (!isAbsolute(value.root) && !win32.isAbsolute(value.root)) ||
      value.url !== undefined
    ) {
      throw new Error(
        "Filesystem publishing requires an absolute root and no URL",
      );
    }
    return { provider: "filesystem", root: value.root };
  }

  if (value.provider === "azure-files") {
    if (typeof value.url !== "string") {
      throw new Error("Azure Files publishing requires a URL");
    }
    const url = new URL(value.url);
    if (
      url.protocol !== "https:" ||
      url.username ||
      url.password ||
      url.search ||
      url.hash ||
      url.pathname === "/" ||
      value.root !== undefined
    ) {
      throw new Error(
        "Azure Files publishing requires an HTTPS share URL without credentials, query parameters or a filesystem root",
      );
    }
    return { provider: "azure-files", url: url.href.replace(/\/$/, "") };
  }

  throw new Error("Unsupported publishing provider in connection file");
}

export function validateConnection(value, expectedServer) {
  knownFields(
    value,
    ["version", "name", "server", "siteBaseURL", "publishing", "capabilities"],
    "Connection file",
  );
  if (value.version !== 1) {
    throw new Error("Unsupported connection file version; expected version 1");
  }
  if (
    typeof value.name !== "string" ||
    !value.name.trim() ||
    /[\x00-\x1f\x7f]/.test(value.name)
  ) {
    throw new Error("Connection file requires a printable platform name");
  }

  const server = platformOrigin(value.server);
  if (expectedServer) {
    const expected = new URL(platformOrigin(expectedServer));
    const actual = new URL(server);
    const sameLocalOrigin =
      isLocalHost(expected.hostname) &&
      isLocalHost(actual.hostname) &&
      expected.port === actual.port &&
      expected.protocol === actual.protocol;
    if (expected.origin !== actual.origin && !sameLocalOrigin) {
      throw new Error("Connection file belongs to a different platform URL");
    }
  }

  const siteBaseURL = platformOrigin(value.siteBaseURL);
  const hostname = new URL(siteBaseURL).hostname;
  if (
    isIP(hostname) ||
    hostname.startsWith("[") ||
    hostname
      .split(".")
      .some((label) => !/^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/.test(label))
  ) {
    throw new Error("siteBaseURL requires a DNS hostname for site subdomains");
  }

  const capabilities = value.capabilities;
  knownFields(
    capabilities,
    ["version", "sites", "files", "database", "realtime", "maxUploadBytes"],
    "Capabilities",
  );
  if (
    capabilities.version !== 1 ||
    ["sites", "files", "database", "realtime"].some(
      (key) => typeof capabilities[key] !== "boolean",
    ) ||
    !Number.isSafeInteger(capabilities.maxUploadBytes) ||
    capabilities.maxUploadBytes < 1
  ) {
    throw new Error("Invalid capability description");
  }

  return {
    version: 1,
    name: value.name.trim(),
    server,
    siteBaseURL,
    publishing: validatePublishing(value.publishing, server),
    capabilities: { ...capabilities },
  };
}
